### ISSUE TITLE
More structured output specifications

### ISSUE THREAD

**Issue by ctmay4** [imsweb/staging-client-java#8]
Schema definitions currently gives detailed information about all `inputs`.  The output that comes back from staging calls are not currently documented in any way.  There are a few issues with this:

1. Users of the staging call may not understand all the outputs that are produced.
2. There are no tables associated with outputs to describe the values.
3. Schemas may produce temporary outputs needed in the staging calculation that do not need to be returned as output.

I think the solution is to add a list of `outputs` to the schema.  Each output would contain:

- key
- name
- table (optional)

There are other possibilities we may want to consider:

- metadata (this might be useful at some point, but I'm not sure)
- agency requirement (we have this for inputs, but I'm not sure we want to try to maintain this on the output side)
- description (some fields may not have a lookup and we might want a longer set of text in addition to the name)

At the end of the staging call only keys listed in the `outputs` would be included in the results.  Some intermediate values might be interesting and could be added to the `outputs` if needed (like the "s" value in the "Testis" schema).

**Comment by schusslern** [comment]
this sounds like a good thing to attempt to do :)

i agree agency requirements may not be worth our time/effort.  the editors don't need to know this, just the vendors and only at the point of screen construction.  I think they should be able to look it up if they don't already know.

I think a description is a good idea - it would be an ideal place to store things like 'Calculation value only,' or something similar, so they don't drive themselves nuts trying to figure out what it is and if anyone wants it.

**Comment by depryf** [comment]
I think that's a good idea. I would add description but wait for the metadata and agency.

**Comment by ctmay4** [comment]
If we add `description` to the `outputs` we should probably also add it to the `inputs` to be consistent.

I definitely think we should wait on adding `agency`, but am a little more on the fence about `metadata`.

**Comment by stealthv** [comment]
It would be nice to have better names for the outputs. What is the timeframe for this? Seems like it will be more than a minor change for Helios to support.

What if an output doesn't exist for all staging paths? Does that need to be indicated somehow?

Also, what if different paths use the same output but it means different things. Not sure that ever happens but it's worth considering.

This will also require more validation to make sure that we don't output values that aren't in the output table.

**Comment by ctmay4** [comment]
There is not current timeframe for this.  I just think it is something we should consider doing.

I was thinking that if not outputs were specified, then all outputs would be included in the results.  That way the CS algorithm would work the same as it does now.  Future algorithms could add output definitions however.

I don't think that a single key will mean two different things depending on the path.  We have no examples of that now and I can't think of one that would ever make sense.  They can define multiple outputs to handle that.

I was thinking if `outputs` were defined and one of the outputs contained values not in the specified table, then I would add an error to the staging results.

**Comment by ctmay4** [comment]
I've done some more work on this and I made one additional change.  I added `default` to the `output` entity.  This is very similar to `initial_context`, but it seems to make more sense to define when you are defining the output.  `initial_context` still has a use.  

1. If you are defining an output and you want it to have a default value, then do it in the `outputs`.
2. If you need a key to have a default for the calculation of stage but do not want that key to be part of the output, add it to `initial_context`.

Still working through all the details and I still have tests to write.  We will discuss before I merge.

**Comment by ctmay4** [comment]
This has been merged and will be released with 1.4.
